src/ray/gcs/gcs_server/gcs_actor_distribution.h

```diff
@@ -0,0 +1,468 @@
+  /// \param job_config Configuration of the job.
+  explicit GcsJobSchedulingContext(const GcsJobConfig &job_config);
+
+  /// Add a actor worker assignment.
+  ///
+  /// \param actor_worker_assignment to be added.
+  /// \return true if the job available resources are enough, else false.
+  bool AddActorWorkerAssignment(
+      std::shared_ptr<GcsActorWorkerAssignment> actor_worker_assignment);
+
+  /// Get available resources of the job.
+  const ResourceSet &GetAvailableResources() const;
+
+  /// Get configuration of the job.
+  const GcsJobConfig &GetJobConfig() const;
+
+  /// Get mutable configuration of the job.
+  GcsJobConfig *GetMutableJobConfig();
+
+  /// Get shared actor worker assignments.
+  const ActorWorkerAssignmentMap &GetSharedActorWorkerAssignments() const;
+
+  /// Get node to actor worker assignments distribution.
+  const NodeToActorWorkerAssignmentsMap &GetNodeToActorWorkerAssignments() const;
+
+  /// Update the node to actor worker assignments distribution.
+  ///
+  /// \param actor_worker_assignment actor worker assignment which node id is updated.
+  bool UpdateNodeToActorWorkerAssignment(
+      std::shared_ptr<GcsActorWorkerAssignment> actor_worker_assignment);
+
+  /// Remove actor worker assignment by the specified node id and actor worker assignment
+  /// id.
+  ///
+  /// The resources of removed actor worker assignment will be released to the job
+  /// scheduling resources.
+  ///
+  /// \param node_id ID of the specified node.
+  /// \param actor_worker_assignment_id ID of the gcs actor worker assignment to be
+  /// removed. \return actor worker assignment associated with the specified node id and
+  /// actor worker assignment id.
+  std::shared_ptr<GcsActorWorkerAssignment>
+  RemoveActorWorkerAssignmentByActorWorkerAssignmentID(
+      const NodeID &node_id, const UniqueID &actor_worker_assignment_id);
+
+  /// Add dummy shared actor worker assignment which does not actually allocate resources
+  /// from the cluster, but only subtracts from the resources declared by the job.
+  bool AddDummySharedActorWorkerAssignment();
+
+  std::shared_ptr<GcsActorWorkerAssignment> GetActorWorkerAssignmentById(
+      const UniqueID &actor_worker_assignment_id) const;
+
+ private:
+  /// Configuration of the job.
+  GcsJobConfig job_config_;
+  /// Shared actor worker assignments.
+  ActorWorkerAssignmentMap shared_actor_worker_assignments_;
+  /// Sole actor worker assignments.
+  ActorWorkerAssignmentMap sole_actor_worker_assignments_;
+  /// Node to worker distribution.
+  NodeToActorWorkerAssignmentsMap node_to_actor_worker_assignments_;
+  /// The job claimed resources when submitting.
+  SchedulingResources scheduling_resources_;
+  uint64_t last_report_time_ms_ = 0;
+};
+
+/// `GcsJobDistribution` represents job distribution on the cluster nodes.
+/// It is responsible for the lifetime of all the `GcsJobSchedulingContext` and records
+/// the distribution of jobs on each node.
+class GcsJobDistribution {
```

Comment:
Is this class store the scheduling result? Could it be named with ...State or ...Store for clarity?

Reply:
This class bookkeeps how jobs are distributed across nodes. It is a status. But 'allocation' seems to be better for some fine-grained actions, such as 'allocating a task to cpu#0'. Any suggestions? @ericl